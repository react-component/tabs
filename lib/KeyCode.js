"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = {
  /**
   * LEFT
   */
  LEFT: 37, // also NUM_WEST
  /**
   * UP
   */
  UP: 38, // also NUM_NORTH
  /**
   * RIGHT
   */
  RIGHT: 39, // also NUM_EAST
  /**
   * DOWN
   */
  DOWN: 40 // also NUM_SOUTH
};
module.exports = exports['default'];