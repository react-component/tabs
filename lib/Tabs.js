'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _KeyCode = require('./KeyCode');

var _KeyCode2 = _interopRequireDefault(_KeyCode);

var _TabPane = require('./TabPane');

var _TabPane2 = _interopRequireDefault(_TabPane);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _utils = require('./utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function noop() {}

function getDefaultActiveKey(props) {
  var activeKey = void 0;
  _react2['default'].Children.forEach(props.children, function (child) {
    if (child && !activeKey && !child.props.disabled) {
      activeKey = child.key;
    }
  });
  return activeKey;
}

function activeKeyIsValid(props, key) {
  var keys = _react2['default'].Children.map(props.children, function (child) {
    return child && child.key;
  });
  return keys.indexOf(key) >= 0;
}

var Tabs = function (_React$Component) {
  (0, _inherits3['default'])(Tabs, _React$Component);

  function Tabs(props) {
    (0, _classCallCheck3['default'])(this, Tabs);

    var _this = (0, _possibleConstructorReturn3['default'])(this, (Tabs.__proto__ || Object.getPrototypeOf(Tabs)).call(this, props));

    _initialiseProps.call(_this);

    var activeKey = void 0;
    if ('activeKey' in props) {
      activeKey = props.activeKey;
    } else if ('defaultActiveKey' in props) {
      activeKey = props.defaultActiveKey;
    } else {
      activeKey = getDefaultActiveKey(props);
    }

    _this.state = {
      activeKey: activeKey
    };
    return _this;
  }

  (0, _createClass3['default'])(Tabs, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if ('activeKey' in nextProps) {
        this.setState({
          activeKey: nextProps.activeKey
        });
      } else if (!activeKeyIsValid(nextProps, this.state.activeKey)) {
        // https://github.com/ant-design/ant-design/issues/7093
        this.setState({
          activeKey: getDefaultActiveKey(nextProps)
        });
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _classnames;

      var props = this.props;
      var prefixCls = props.prefixCls,
          navWrapper = props.navWrapper,
          tabBarPosition = props.tabBarPosition,
          className = props.className,
          renderTabContent = props.renderTabContent,
          renderTabBar = props.renderTabBar,
          destroyInactiveTabPane = props.destroyInactiveTabPane,
          restProps = (0, _objectWithoutProperties3['default'])(props, ['prefixCls', 'navWrapper', 'tabBarPosition', 'className', 'renderTabContent', 'renderTabBar', 'destroyInactiveTabPane']);

      var cls = (0, _classnames3['default'])((_classnames = {}, (0, _defineProperty3['default'])(_classnames, prefixCls, 1), (0, _defineProperty3['default'])(_classnames, prefixCls + '-' + tabBarPosition, 1), (0, _defineProperty3['default'])(_classnames, className, !!className), _classnames));

      this.tabBar = renderTabBar();
      var contents = [_react2['default'].cloneElement(this.tabBar, {
        prefixCls: prefixCls,
        navWrapper: navWrapper,
        key: 'tabBar',
        onKeyDown: this.onNavKeyDown,
        tabBarPosition: tabBarPosition,
        onTabClick: this.onTabClick,
        panels: props.children,
        activeKey: this.state.activeKey
      }), _react2['default'].cloneElement(renderTabContent(), {
        prefixCls: prefixCls,
        tabBarPosition: tabBarPosition,
        activeKey: this.state.activeKey,
        destroyInactiveTabPane: destroyInactiveTabPane,
        children: props.children,
        onChange: this.setActiveKey,
        key: 'tabContent'
      })];
      if (tabBarPosition === 'bottom') {
        contents.reverse();
      }
      return _react2['default'].createElement(
        'div',
        (0, _extends3['default'])({
          className: cls,
          style: props.style
        }, (0, _utils.getDataAttr)(restProps)),
        contents
      );
    }
  }]);
  return Tabs;
}(_react2['default'].Component);

var _initialiseProps = function _initialiseProps() {
  var _this2 = this;

  this.onTabClick = function (activeKey, e) {
    if (_this2.tabBar.props.onTabClick) {
      _this2.tabBar.props.onTabClick(activeKey, e);
    }
    _this2.setActiveKey(activeKey);
  };

  this.onNavKeyDown = function (e) {
    var canDownNavKey = _this2.props.canDownNavKey;


    if (canDownNavKey && !canDownNavKey()) {
      return;
    }

    var eventKeyCode = e.keyCode;
    if (eventKeyCode === _KeyCode2['default'].RIGHT || eventKeyCode === _KeyCode2['default'].DOWN) {
      e.preventDefault();
      var nextKey = _this2.getNextActiveKey(true);
      _this2.onTabClick(nextKey);
    } else if (eventKeyCode === _KeyCode2['default'].LEFT || eventKeyCode === _KeyCode2['default'].UP) {
      e.preventDefault();
      var previousKey = _this2.getNextActiveKey(false);
      _this2.onTabClick(previousKey);
    }
  };

  this.setActiveKey = function (activeKey) {
    if (_this2.state.activeKey !== activeKey) {
      if (!('activeKey' in _this2.props)) {
        _this2.setState({
          activeKey: activeKey
        });
      }
      _this2.props.onChange(activeKey);
    }
  };

  this.getNextActiveKey = function (next) {
    var activeKey = _this2.state.activeKey;
    var children = [];
    _react2['default'].Children.forEach(_this2.props.children, function (c) {
      if (c && !c.props.disabled) {
        if (next) {
          children.push(c);
        } else {
          children.unshift(c);
        }
      }
    });
    var length = children.length;
    var ret = length && children[0].key;
    children.forEach(function (child, i) {
      if (child.key === activeKey) {
        if (i === length - 1) {
          ret = children[0].key;
        } else {
          ret = children[i + 1].key;
        }
      }
    });
    return ret;
  };
};

exports['default'] = Tabs;


Tabs.propTypes = {
  destroyInactiveTabPane: _propTypes2['default'].bool,
  renderTabBar: _propTypes2['default'].func.isRequired,
  renderTabContent: _propTypes2['default'].func.isRequired,
  navWrapper: _propTypes2['default'].func,
  onChange: _propTypes2['default'].func,
  children: _propTypes2['default'].any,
  prefixCls: _propTypes2['default'].string,
  className: _propTypes2['default'].string,
  tabBarPosition: _propTypes2['default'].string,
  style: _propTypes2['default'].object,
  activeKey: _propTypes2['default'].string,
  defaultActiveKey: _propTypes2['default'].string,
  canDownNavKey: _propTypes2['default'].func
};

Tabs.defaultProps = {
  prefixCls: 'rc-tabs',
  destroyInactiveTabPane: false,
  onChange: noop,
  navWrapper: function navWrapper(arg) {
    return arg;
  },
  tabBarPosition: 'top',
  style: {}
};

Tabs.TabPane = _TabPane2['default'];
module.exports = exports['default'];