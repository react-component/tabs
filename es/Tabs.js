import _extends from 'babel-runtime/helpers/extends';
import _defineProperty from 'babel-runtime/helpers/defineProperty';
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import React from 'react';
import PropTypes from 'prop-types';
import KeyCode from './KeyCode';
import TabPane from './TabPane';
import classnames from 'classnames';
import { getDataAttr } from './utils';

function noop() {}

function getDefaultActiveKey(props) {
  var activeKey = void 0;
  React.Children.forEach(props.children, function (child) {
    if (child && !activeKey && !child.props.disabled) {
      activeKey = child.key;
    }
  });
  return activeKey;
}

function activeKeyIsValid(props, key) {
  var keys = React.Children.map(props.children, function (child) {
    return child && child.key;
  });
  return keys.indexOf(key) >= 0;
}

var Tabs = function (_React$Component) {
  _inherits(Tabs, _React$Component);

  function Tabs(props) {
    _classCallCheck(this, Tabs);

    var _this = _possibleConstructorReturn(this, (Tabs.__proto__ || Object.getPrototypeOf(Tabs)).call(this, props));

    _initialiseProps.call(_this);

    var activeKey = void 0;
    if ('activeKey' in props) {
      activeKey = props.activeKey;
    } else if ('defaultActiveKey' in props) {
      activeKey = props.defaultActiveKey;
    } else {
      activeKey = getDefaultActiveKey(props);
    }

    _this.state = {
      activeKey: activeKey
    };
    return _this;
  }

  _createClass(Tabs, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if ('activeKey' in nextProps) {
        this.setState({
          activeKey: nextProps.activeKey
        });
      } else if (!activeKeyIsValid(nextProps, this.state.activeKey)) {
        // https://github.com/ant-design/ant-design/issues/7093
        this.setState({
          activeKey: getDefaultActiveKey(nextProps)
        });
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _classnames;

      var props = this.props;

      var prefixCls = props.prefixCls,
          navWrapper = props.navWrapper,
          tabBarPosition = props.tabBarPosition,
          className = props.className,
          renderTabContent = props.renderTabContent,
          renderTabBar = props.renderTabBar,
          destroyInactiveTabPane = props.destroyInactiveTabPane,
          restProps = _objectWithoutProperties(props, ['prefixCls', 'navWrapper', 'tabBarPosition', 'className', 'renderTabContent', 'renderTabBar', 'destroyInactiveTabPane']);

      var cls = classnames((_classnames = {}, _defineProperty(_classnames, prefixCls, 1), _defineProperty(_classnames, prefixCls + '-' + tabBarPosition, 1), _defineProperty(_classnames, className, !!className), _classnames));

      this.tabBar = renderTabBar();
      var contents = [React.cloneElement(this.tabBar, {
        prefixCls: prefixCls,
        navWrapper: navWrapper,
        key: 'tabBar',
        onKeyDown: this.onNavKeyDown,
        tabBarPosition: tabBarPosition,
        onTabClick: this.onTabClick,
        panels: props.children,
        activeKey: this.state.activeKey
      }), React.cloneElement(renderTabContent(), {
        prefixCls: prefixCls,
        tabBarPosition: tabBarPosition,
        activeKey: this.state.activeKey,
        destroyInactiveTabPane: destroyInactiveTabPane,
        children: props.children,
        onChange: this.setActiveKey,
        key: 'tabContent'
      })];
      if (tabBarPosition === 'bottom') {
        contents.reverse();
      }
      return React.createElement(
        'div',
        _extends({
          className: cls,
          style: props.style
        }, getDataAttr(restProps)),
        contents
      );
    }
  }]);

  return Tabs;
}(React.Component);

var _initialiseProps = function _initialiseProps() {
  var _this2 = this;

  this.onTabClick = function (activeKey, e) {
    if (_this2.tabBar.props.onTabClick) {
      _this2.tabBar.props.onTabClick(activeKey, e);
    }
    _this2.setActiveKey(activeKey);
  };

  this.onNavKeyDown = function (e) {
    var canDownNavKey = _this2.props.canDownNavKey;


    if (canDownNavKey && !canDownNavKey()) {
      return;
    }

    var eventKeyCode = e.keyCode;
    if (eventKeyCode === KeyCode.RIGHT || eventKeyCode === KeyCode.DOWN) {
      e.preventDefault();
      var nextKey = _this2.getNextActiveKey(true);
      _this2.onTabClick(nextKey);
    } else if (eventKeyCode === KeyCode.LEFT || eventKeyCode === KeyCode.UP) {
      e.preventDefault();
      var previousKey = _this2.getNextActiveKey(false);
      _this2.onTabClick(previousKey);
    }
  };

  this.setActiveKey = function (activeKey) {
    if (_this2.state.activeKey !== activeKey) {
      if (!('activeKey' in _this2.props)) {
        _this2.setState({
          activeKey: activeKey
        });
      }
      _this2.props.onChange(activeKey);
    }
  };

  this.getNextActiveKey = function (next) {
    var activeKey = _this2.state.activeKey;
    var children = [];
    React.Children.forEach(_this2.props.children, function (c) {
      if (c && !c.props.disabled) {
        if (next) {
          children.push(c);
        } else {
          children.unshift(c);
        }
      }
    });
    var length = children.length;
    var ret = length && children[0].key;
    children.forEach(function (child, i) {
      if (child.key === activeKey) {
        if (i === length - 1) {
          ret = children[0].key;
        } else {
          ret = children[i + 1].key;
        }
      }
    });
    return ret;
  };
};

export default Tabs;


Tabs.propTypes = {
  destroyInactiveTabPane: PropTypes.bool,
  renderTabBar: PropTypes.func.isRequired,
  renderTabContent: PropTypes.func.isRequired,
  navWrapper: PropTypes.func,
  onChange: PropTypes.func,
  children: PropTypes.any,
  prefixCls: PropTypes.string,
  className: PropTypes.string,
  tabBarPosition: PropTypes.string,
  style: PropTypes.object,
  activeKey: PropTypes.string,
  defaultActiveKey: PropTypes.string,
  canDownNavKey: PropTypes.func
};

Tabs.defaultProps = {
  prefixCls: 'rc-tabs',
  destroyInactiveTabPane: false,
  onChange: noop,
  navWrapper: function navWrapper(arg) {
    return arg;
  },
  tabBarPosition: 'top',
  style: {}
};

Tabs.TabPane = TabPane;