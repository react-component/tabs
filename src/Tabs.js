import React from 'react';
import PropTypes from 'prop-types';
import KeyCode from './KeyCode';
import TabPane from './TabPane';
import classnames from 'classnames';
import { getDataAttr } from './utils';

function noop() {
}

function getDefaultActiveKey(props) {
  let activeKey;
  React.Children.forEach(props.children, (child) => {
    if (child && !activeKey && !child.props.disabled) {
      activeKey = child.key;
    }
  });
  return activeKey;
}

function activeKeyIsValid(props, key) {
  const keys = React.Children.map(props.children, child => child && child.key);
  return keys.indexOf(key) >= 0;
}

export default class Tabs extends React.Component {
  constructor(props) {
    super(props);

    let activeKey;
    if ('activeKey' in props) {
      activeKey = props.activeKey;
    } else if ('defaultActiveKey' in props) {
      activeKey = props.defaultActiveKey;
    } else {
      activeKey = getDefaultActiveKey(props);
    }

    this.state = {
      activeKey,
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('activeKey' in nextProps) {
      this.setState({
        activeKey: nextProps.activeKey,
      });
    } else if (!activeKeyIsValid(nextProps, this.state.activeKey)) {
      // https://github.com/ant-design/ant-design/issues/7093
      this.setState({
        activeKey: getDefaultActiveKey(nextProps),
      });
    }
  }

  onTabClick = (activeKey, e) => {
    if (this.tabBar.props.onTabClick) {
      this.tabBar.props.onTabClick(activeKey, e);
    }
    this.setActiveKey(activeKey);
  }

  onNavKeyDown = (e) => {
    const { canDownNavKey } = this.props;

    if (canDownNavKey && !canDownNavKey()) {
      return;
    }

    const eventKeyCode = e.keyCode;
    if (eventKeyCode === KeyCode.RIGHT || eventKeyCode === KeyCode.DOWN) {
      e.preventDefault();
      const nextKey = this.getNextActiveKey(true);
      this.onTabClick(nextKey);
    } else if (eventKeyCode === KeyCode.LEFT || eventKeyCode === KeyCode.UP) {
      e.preventDefault();
      const previousKey = this.getNextActiveKey(false);
      this.onTabClick(previousKey);
    }
  }

  setActiveKey = (activeKey) => {
    if (this.state.activeKey !== activeKey) {
      if (!('activeKey' in this.props)) {
        this.setState({
          activeKey,
        });
      }
      this.props.onChange(activeKey);
    }
  }

  getNextActiveKey = (next) => {
    const activeKey = this.state.activeKey;
    const children = [];
    React.Children.forEach(this.props.children, (c) => {
      if (c && !c.props.disabled) {
        if (next) {
          children.push(c);
        } else {
          children.unshift(c);
        }
      }
    });
    const length = children.length;
    let ret = length && children[0].key;
    children.forEach((child, i) => {
      if (child.key === activeKey) {
        if (i === length - 1) {
          ret = children[0].key;
        } else {
          ret = children[i + 1].key;
        }
      }
    });
    return ret;
  }

  render() {
    const props = this.props;
    const {
      prefixCls,
      navWrapper,
      tabBarPosition, className,
      renderTabContent,
      renderTabBar,
      destroyInactiveTabPane,
      ...restProps,
    } = props;
    const cls = classnames({
      [prefixCls]: 1,
      [`${prefixCls}-${tabBarPosition}`]: 1,
      [className]: !!className,
    });

    this.tabBar = renderTabBar();
    const contents = [
      React.cloneElement(this.tabBar, {
        prefixCls,
        navWrapper,
        key: 'tabBar',
        onKeyDown: this.onNavKeyDown,
        tabBarPosition,
        onTabClick: this.onTabClick,
        panels: props.children,
        activeKey: this.state.activeKey,
      }),
      React.cloneElement(renderTabContent(), {
        prefixCls,
        tabBarPosition,
        activeKey: this.state.activeKey,
        destroyInactiveTabPane,
        children: props.children,
        onChange: this.setActiveKey,
        key: 'tabContent',
      }),
    ];
    if (tabBarPosition === 'bottom') {
      contents.reverse();
    }
    return (
      <div
        className={cls}
        style={props.style}
        {...getDataAttr(restProps)}
      >
        {contents}
      </div>
    );
  }
}

Tabs.propTypes = {
  destroyInactiveTabPane: PropTypes.bool,
  renderTabBar: PropTypes.func.isRequired,
  renderTabContent: PropTypes.func.isRequired,
  navWrapper: PropTypes.func,
  onChange: PropTypes.func,
  children: PropTypes.any,
  prefixCls: PropTypes.string,
  className: PropTypes.string,
  tabBarPosition: PropTypes.string,
  style: PropTypes.object,
  activeKey: PropTypes.string,
  defaultActiveKey: PropTypes.string,
  canDownNavKey: PropTypes.func,
};

Tabs.defaultProps = {
  prefixCls: 'rc-tabs',
  destroyInactiveTabPane: false,
  onChange: noop,
  navWrapper: arg => arg,
  tabBarPosition: 'top',
  style: {},
};

Tabs.TabPane = TabPane;
