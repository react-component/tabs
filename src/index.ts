import Tabs, { TabsProps } from './Tabs';
import TabPane, { TabPaneProps } from './TabPanelList/TabPane';

export { TabPane, TabsProps, TabPaneProps };

export default Tabs;
