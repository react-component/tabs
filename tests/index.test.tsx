import React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import KeyCode from 'rc-util/lib/KeyCode';
import Tabs, { TabPane } from '../src';
import { TabsProps } from '../src/Tabs';

describe('Tabs.Basic', () => {
  function getTabs(props: TabsProps = null) {
    const mergedProps = {
      children: [
        <TabPane tab="light" key="light">
          Light
        </TabPane>,
        <TabPane tab="bamboo" key="bamboo">
          Bamboo
        </TabPane>,
        <TabPane tab="cute" key="cute">
          Cute
        </TabPane>,
      ],
      ...props,
    };

    return <Tabs {...mergedProps} />;
  }

  it('Normal', () => {
    const wrapper = mount(getTabs({ defaultActiveKey: 'bamboo' }));

    expect(wrapper.render()).toMatchSnapshot();
  });

  it('Skip invalidate children', () => {
    const wrapper = mount(
      getTabs({
        children: [
          <TabPane tab="light" key="light">
            Light
          </TabPane>,
          'not me',
        ],
      }),
    );
    wrapper.update();

    expect(wrapper.render()).toMatchSnapshot();
  });

  it('nothing for empty tabs', () => {
    mount(getTabs({ children: null }));
  });

  describe('onChange and onTabClick should work', () => {
    const list: { name: string; trigger: (wrapper: ReactWrapper) => void }[] = [
      {
        name: 'outer div',
        trigger: wrapper =>
          wrapper
            .find('.rc-tabs-tab')
            .at(2)
            .simulate('click'),
      },
      {
        name: 'inner button',
        trigger: wrapper =>
          wrapper
            .find('.rc-tabs-tab .rc-tabs-tab-btn')
            .at(2)
            .simulate('click'),
      },
      {
        name: 'inner button key down',
        trigger: wrapper =>
          wrapper
            .find('.rc-tabs-tab .rc-tabs-tab-btn')
            .at(2)
            .simulate('keydown', {
              which: KeyCode.SPACE,
            }),
      },
    ];

    list.forEach(({ name, trigger }) => {
      it(name, () => {
        const onChange = jest.fn();
        const onTabClick = jest.fn();
        const wrapper = mount(getTabs({ onChange, onTabClick }));

        trigger(wrapper);
        expect(onTabClick).toHaveBeenCalledWith('cute', expect.anything());
        expect(onChange).toHaveBeenCalledWith('cute');
      });
    });
  });

  it('active first tab when children is changed', () => {
    const wrapper = mount(getTabs());
    wrapper.setProps({
      children: (
        <TabPane tab="Yo" key="2333">
          New
        </TabPane>
      ),
    });
    wrapper.update();

    expect(wrapper.find('.rc-tabs-tab-active').text()).toEqual('Yo');
  });

  it('active first tab when children is not changed at controlled mode', () => {
    const wrapper = mount(getTabs({ activeKey: 'light' }));
    expect(wrapper.find('.rc-tabs-tab-active').text()).toEqual('light');

    wrapper.setProps({
      children: (
        <TabPane tab="Yo" key="2333">
          New
        </TabPane>
      ),
    });
    expect(wrapper.find('.rc-tabs-tab-active')).toHaveLength(0);
  });

  it('tabBarGutter should work', () => {
    const topTabs = mount(getTabs({ tabBarGutter: 23 }));
    expect(
      topTabs
        .find('.rc-tabs-tab')
        .first()
        .props().style.marginRight,
    ).toEqual(23);

    const rightTabs = mount(getTabs({ tabBarGutter: 33, tabPosition: 'right' }));
    expect(
      rightTabs
        .find('.rc-tabs-tab')
        .first()
        .props().style.marginBottom,
    ).toEqual(33);
  });

  describe('renderTabBar', () => {
    it('works', () => {
      const renderTabBar = jest.fn((props, Component) => {
        return (
          <div className="my-wrapper">
            <Component {...props}>{node => <span className="my-node">{node}</span>}</Component>
          </div>
        );
      });
      const wrapper = mount(getTabs({ renderTabBar }));
      expect(wrapper.find('.my-wrapper').length).toBeTruthy();
      expect(wrapper.find('.my-node').length).toBeTruthy();
      expect(renderTabBar).toHaveBeenCalled();
    });
    it('has panes property in props', () => {
      const renderTabBar = props => {
        return (
          <div>
            {props.panes.map(pane => (
              <span key={pane.key} data-key={pane.key}>
                tab
              </span>
            ))}
          </div>
        );
      };
      const wrapper = mount(getTabs({ renderTabBar }));
      expect(wrapper.find('[data-key="light"]').length).toBeTruthy();
      expect(wrapper.find('[data-key="bamboo"]').length).toBeTruthy();
      expect(wrapper.find('[data-key="cute"]').length).toBeTruthy();
    });
  });

  it('destroyInactiveTabPane', () => {
    const wrapper = mount(
      getTabs({
        activeKey: 'light',
        destroyInactiveTabPane: true,
        children: [<TabPane key="light">Light</TabPane>, <TabPane key="bamboo">Bamboo</TabPane>],
      }),
    );

    function matchText(light: string, bamboo: string) {
      expect(
        wrapper
          .find('.rc-tabs-tabpane')
          .first()
          .text(),
      ).toEqual(light);
      expect(
        wrapper
          .find('.rc-tabs-tabpane')
          .last()
          .text(),
      ).toEqual(bamboo);
    }

    matchText('Light', '');

    wrapper.setProps({ activeKey: 'bamboo' });
    matchText('', 'Bamboo');
  });

  describe('editable', () => {
    it('no and', () => {
      const onEdit = jest.fn();
      const wrapper = mount(getTabs({ editable: { onEdit, showAdd: false } }));
      expect(wrapper.find('.rc-tabs-nav-add')).toHaveLength(0);
    });

    it('add', () => {
      const onEdit = jest.fn();
      const wrapper = mount(getTabs({ editable: { onEdit } }));
      wrapper
        .find('.rc-tabs-nav-add')
        .first()
        .simulate('click');
      expect(onEdit).toHaveBeenCalledWith('add', {
        key: undefined,
        event: expect.anything(),
      });
    });

    const list: { name: string; trigger: (node: ReactWrapper) => void }[] = [
      {
        name: 'click',
        trigger: node => {
          node.simulate('click');
        },
      },
    ];

    list.forEach(({ name, trigger }) => {
      it(`remove by ${name}`, () => {
        const onEdit = jest.fn();
        const wrapper = mount(getTabs({ editable: { onEdit } }));

        const first = wrapper.find('.rc-tabs-tab-remove').first();
        trigger(first);

        // Should be button to enable press SPACE key to trigger
        expect(first.instance() instanceof HTMLButtonElement).toBeTruthy();

        expect(onEdit).toHaveBeenCalledWith('remove', {
          key: 'light',
          event: expect.anything(),
        });
      });
    });

    it('customize closeIcon', () => {
      const onEdit = jest.fn();
      const wrapper = mount(
        getTabs({
          editable: { onEdit },
          children: [
            <TabPane tab="light" closeIcon={<span className="close-light" />}>
              Light
            </TabPane>,
          ],
        }),
      );

      expect(wrapper.find('.rc-tabs-tab-remove').find('.close-light').length).toBeTruthy();
    });
  });

  it('extra', () => {
    const wrapper = mount(getTabs({ tabBarExtraContent: 'Bamboo' }));
    expect(wrapper.find('.rc-tabs-extra-content').text()).toEqual('Bamboo');
  });

  it('extra positon', () => {
    const wrapper = mount(
      getTabs({ tabBarExtraContent: { left: 'Left Bamboo', right: 'Right Bamboo' } }),
    );
    expect(
      wrapper
        .find('.rc-tabs-extra-content')
        .first()
        .text(),
    ).toEqual('Left Bamboo');

    expect(
      wrapper
        .find('.rc-tabs-extra-content')
        .at(1)
        .text(),
    ).toEqual('Right Bamboo');
  });

  describe('animated', () => {
    it('false', () => {
      const wrapper = mount(getTabs({ animated: false }));
      expect(wrapper.find('TabPanelList').prop('animated')).toEqual({
        inkBar: false,
        tabPane: false,
      });
    });

    it('true', () => {
      const wrapper = mount(getTabs({ animated: true }));
      expect(wrapper.find('TabPanelList').prop('animated')).toEqual({
        inkBar: true,
        tabPane: false,
      });
    });

    it('customize', () => {
      const wrapper = mount(getTabs({ animated: { inkBar: false, tabPane: true } }));
      expect(wrapper.find('TabPanelList').prop('animated')).toEqual({
        inkBar: false,
        tabPane: true,
      });
    });
  });

  it('focus to scroll', () => {
    const wrapper = mount(getTabs());
    wrapper
      .find('.rc-tabs-tab')
      .first()
      .simulate('focus');

    wrapper.unmount();
  });

  it('tabBarStyle', () => {
    const wrapper = mount(getTabs({ tabBarStyle: { background: 'red' } }));
    expect(wrapper.find('.rc-tabs-nav').prop('style').background).toEqual('red');
  });
});
