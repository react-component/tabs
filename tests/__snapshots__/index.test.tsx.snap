// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs.Basic Normal 1`] = `
<div
  class="rc-tabs rc-tabs-top"
>
  <div
    class="rc-tabs-nav"
    role="tablist"
  >
    <div
      class="rc-tabs-nav-wrap"
    >
      <div
        class="rc-tabs-nav-list"
        style="transform: translate(0px, 0px);"
      >
        <div
          class="rc-tabs-tab"
        >
          <div
            aria-controls="rc-tabs-test-panel-light"
            aria-selected="false"
            class="rc-tabs-tab-btn"
            id="rc-tabs-test-tab-light"
            role="tab"
            tabindex="0"
          >
            light
          </div>
        </div>
        <div
          class="rc-tabs-tab rc-tabs-tab-active"
        >
          <div
            aria-controls="rc-tabs-test-panel-bamboo"
            aria-selected="true"
            class="rc-tabs-tab-btn"
            id="rc-tabs-test-tab-bamboo"
            role="tab"
            tabindex="0"
          >
            bamboo
          </div>
        </div>
        <div
          class="rc-tabs-tab"
        >
          <div
            aria-controls="rc-tabs-test-panel-cute"
            aria-selected="false"
            class="rc-tabs-tab-btn"
            id="rc-tabs-test-tab-cute"
            role="tab"
            tabindex="0"
          >
            cute
          </div>
        </div>
        <div
          class="rc-tabs-ink-bar rc-tabs-ink-bar-animated"
        />
      </div>
    </div>
    <div
      class="rc-tabs-nav-operations rc-tabs-nav-operations-hidden"
    >
      <button
        aria-controls="rc-tabs-test-more-popup"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-hidden="true"
        class="rc-tabs-nav-more"
        id="rc-tabs-test-more"
        style="visibility: hidden; order: 1;"
        tabindex="-1"
        type="button"
      >
        More
      </button>
    </div>
  </div>
  <div
    class="rc-tabs-content-holder"
  >
    <div
      class="rc-tabs-content rc-tabs-content-top"
    >
      <div
        aria-hidden="true"
        aria-labelledby="rc-tabs-test-tab-light"
        class="rc-tabs-tabpane"
        id="rc-tabs-test-panel-light"
        role="tabpanel"
        style="display: none;"
        tabindex="-1"
      />
      <div
        aria-hidden="false"
        aria-labelledby="rc-tabs-test-tab-bamboo"
        class="rc-tabs-tabpane rc-tabs-tabpane-active"
        id="rc-tabs-test-panel-bamboo"
        role="tabpanel"
        tabindex="0"
      >
        Bamboo
      </div>
      <div
        aria-hidden="true"
        aria-labelledby="rc-tabs-test-tab-cute"
        class="rc-tabs-tabpane"
        id="rc-tabs-test-panel-cute"
        role="tabpanel"
        style="display: none;"
        tabindex="-1"
      />
    </div>
  </div>
</div>
`;

exports[`Tabs.Basic Skip invalidate children 1`] = `
<div
  class="rc-tabs rc-tabs-top"
>
  <div
    class="rc-tabs-nav"
    role="tablist"
  >
    <div
      class="rc-tabs-nav-wrap"
    >
      <div
        class="rc-tabs-nav-list"
        style="transform: translate(0px, 0px);"
      >
        <div
          class="rc-tabs-tab rc-tabs-tab-active"
        >
          <div
            aria-controls="rc-tabs-test-panel-light"
            aria-selected="true"
            class="rc-tabs-tab-btn"
            id="rc-tabs-test-tab-light"
            role="tab"
            tabindex="0"
          >
            light
          </div>
        </div>
        <div
          class="rc-tabs-ink-bar rc-tabs-ink-bar-animated"
        />
      </div>
    </div>
    <div
      class="rc-tabs-nav-operations rc-tabs-nav-operations-hidden"
    >
      <button
        aria-controls="rc-tabs-test-more-popup"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-hidden="true"
        class="rc-tabs-nav-more"
        id="rc-tabs-test-more"
        style="visibility: hidden; order: 1;"
        tabindex="-1"
        type="button"
      >
        More
      </button>
    </div>
  </div>
  <div
    class="rc-tabs-content-holder"
  >
    <div
      class="rc-tabs-content rc-tabs-content-top"
    >
      <div
        aria-hidden="false"
        aria-labelledby="rc-tabs-test-tab-light"
        class="rc-tabs-tabpane rc-tabs-tabpane-active"
        id="rc-tabs-test-panel-light"
        role="tabpanel"
        tabindex="0"
      >
        Light
      </div>
    </div>
  </div>
</div>
`;
